metron-analytics/metron-statistics/README.md

```diff
@@ -53,6 +53,32 @@ functions can be used from everywhere where Stellar is used.
   * bounds - A list of value bounds (excluding min and max) in sorted order.
 * Returns: Which bin N the value falls in such that bound(N-1) < value <= bound(N).  No min and max bounds are provided, so values smaller than the 0'th bound go in the 0'th bin, and values greater than the last bound go in the M'th bin.
 
+### Sampling Functions
+
+#### `SAMPLE_ADD`
+* Description: Add a value or collection of values to a sampler.
+* Input:
```

Comment:
This makes it seem like Uniform sampler is a 'known' thing.  But it is not, either by explanation or reference to where it is explained ( as we have done referring to algorithms before ).
Is there another type of sampler?

Somewhere ( I'm not sure where ) we should say:
"A sampler is a xxxxx that is | does | acts as  xxxxx for the sample functions. The default has these properties, but you can override that in init"

Why even mention the Universal?

Reply:
Sorry, `uniform` here is intended to mean that there's each element has equal probability of being in the sample (e.g. the probability is pulled from a [uniform probability distribution](https://en.wikipedia.org/wiki/Uniform_distribution_(continuous))).  I can probably do a better job documenting.

Reply:
Couldn't this be simplified to

```java
 if(ret == null ) {
      if(obj != null) {
         throw new IllegalStateException(argName + "argument(" + obj
                                        + " is expected to be an " + expectedClazz.getName()
                                        + ", but was " + obj
                                        );
       }
     }
return Optional.ofNullable(ret);


```

Reply:
Ok, It seemed like the Uniform implementation was leaking

Reply:
There are definitely other types of reservoir samplers which we will probably want.  Most specifically a sampler that is biased toward recency (so non-uniform in that case).

Reply:
Recency would surely be more relevant for merged resampling in a profile context?

Reply:
They're both needed.  Some use-cases would be fine without bias and some would be better with bias.  As a follow-on, I was planning on adding a biased sampler, but this is a big enough PR without it.  It'd look something like:
```
samples := SAMPLE_MERGE(PROFILE_GET('samples', ...))
biased_sample := SAMPLE_GET_BIASED(samples, 0.015)
```

Reply:
Then we'll have a get sample types method, like we do with other things like this right?

Reply:
Actually, more than likely it'd be a separate init since each type are going to have different types of parameters depending on the algorithm.  So a biased sampler would be `SAMPLE_INIT_BIASED(size, ...)`